tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,396 @@
+#include <deque>
+#include <iostream>
+#include <set>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "llvm/Demangle/Demangle.h"
+#include "llvm/IR/CallSite.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/raw_ostream.h"
+using namespace llvm;
+
+enum OutputFormatType { Dot, YAML };
+static cl::opt<OutputFormatType> OutputFormat(
+    "format",
+    cl::desc("Output format."),
+    cl::values(clEnumValN(Dot, "dot", "print as dot"),
+               clEnumValN(YAML, "yaml", "print as yaml")));
+
+static cl::opt<bool> TransitiveClosure(
+    "closure",
+    cl::desc("Output transitive closure."),
+    cl::init(true));
+
+static cl::opt<int> Verbose(
+    "v",
+    cl::desc("Verbose level"),
+    cl::Hidden,
+    cl::init(0));
+
+static cl::list<std::string> DebugFilters(
+    "df",
+    cl::desc("Debug filter patterns. Example: -df CPUType,at::native"),
+    cl::ZeroOrMore,
+    cl::CommaSeparated);
+
+namespace {
+
+typedef std::set<std::string> SET;
+typedef std::unordered_map<std::string, std::set<std::string>> GRAPH;
+
+// Referenced the logic in llvm-cxxfilt.cpp.
+std::string demangle(const std::string& mangled) {
+  int status;
+  const char* decorated = mangled.c_str();
+  size_t decoratedLength = strlen(decorated);
+
+  char *undecorated = itaniumDemangle(decorated, nullptr, nullptr, &status);
+
+  if (!undecorated &&
+      (decoratedLength > 6 && strncmp(decorated, "__imp_", 6) == 0)) {
+    undecorated = itaniumDemangle(decorated + 6, nullptr, nullptr, &status);
+  }
+  std::string result(undecorated ? undecorated : mangled);
+  free(undecorated);
+  return result;
+}
+
+// LLVM_DEBUG needs opt to be built with debug support.
+template<
+    typename T,
+    typename std::enable_if<std::is_base_of<Value, T>::value, int>::type = 0>
+std::ostream& operator<<(std::ostream& out, T& I) {
+  std::string str;
+  raw_string_ostream O(str);
+  O << I;
+  return out << str;
+}
+
+class OpDependency : public ModulePass {
+public:
+  static char ID; // Pass identification, replacement for typeid
+
+  OpDependency() : ModulePass(ID) {}
+  ~OpDependency() = default;
+
+  bool runOnModule(Module& M) override {
+    // Key nodes are the nodes we keep in the result graph. They are global
+    // strings that match function schema pattern. Function names that match
+    // DebugFilters are also counted as key nodes.
```

Comment:
Is "key" a funny way of spelling "root"? :o) (this is me with my GC hat on lol)

Reply:
Or is this the transitive closure?

Reply:
"key" is usually the roots (all function schema strings starting with "aten::"), but I added a "debug filter" flag to keep intermediate symbols in the result graph for me to verify it doesn't miss important symbols. For example, without debug symbol it outputs:
"aten::abs -> aten::empty"
If I added debug filter "CPUType", it will output:
"aten::abs -> at::CPUType::{anonymous namespace}::aten(...) -> ... -> aten::empty"

This is a convenient way for me to verify all CPUType methods are successfully associated to "aten::..." function schema strings by the fuzzy matcher.

I call these intermediate symbols to be kept in the result graph "key nodes". I call the transitive closure method "simplifyGraph", which means "removing non-key nodes while not breaking transitive connectivity relationship among key nodes".
But I'm not very sensitive to english naming, please let me know if "root" or something else is better for this.

Reply:
I guess your point is, at::CPUType::...::abs() is a reachable node, but it is not a key node (because it is not a point where we can prune code out). I guess key is fine for that, in this case. I think what you said about "intermediate symbols" helped make this clear for me.

Reply:
yeah, maybe I should define these concepts more clear. There are two type of nodes in my graph:
1) pure string constants in source files, e.g. "aten::abs(Tensor) -> Tensor", which represents function schemas;
2) Function / symbols in object files, e.g. "at::CPUType::...::abs()";

When I add them into my graph, I convert them into basic_string representation and call them all as "nodes" since the BFS doesn't need to know the difference.

But ultimately we care about #1 as that's what we use to prune registered ops via codegen.
#2 is kept for debugging purpose.

I'll add these to the comments.